### PR TITLE
Representation arithmetic

### PR DESCRIPTION
**Pull request by mhvk**
This aims to address #5255, by making it possible to do arithmetic on representations, such as addition of two representations, multiplication with a ``Quantity``, or calculating the norm via ``abs``. It also introduces news methods ``mean``, ``sum``, ``dot``, and ``cross`` with obvious meaning.

cc @StuartLittlefair, @eteq, @adrn, @astrofrog

### CODE REVIEW

File: astropy/coordinates/representation.py [hunk 1]
```diff
@@ -221,6 +227,136 @@ def _unitstr(self):
...
+                                 for component in self.components))
+
+    def __ne__(self, other):
+        return np.logical_not(self.__eq__(other))
+
+    def __mul__(self, other):
+        values = (getattr(self, component) for component in self.components)
+        return self.__class__(*(
+            (value if issubclass(cls, Angle) else value * other)
+            for (component, cls), value in zip(self.attr_classes.items(),
```
Comment (MSeifert04) [inline astropy/coordinates/representation.py]: This is really hard to follow. Some questions:

- Why extract the `component` if it's not used in the comprehension? Why not iterate over the `values` directly?
- `zip` could be imported from `extern.six.moves`.
- The multiplication and division only differ by the `value {op} other`, maybe these could be unified?

File: astropy/coordinates/representation.py [hunk 2]
```diff
@@ -221,6 +227,136 @@ def _unitstr(self):
...
+
+    def __abs__(self):
+        return np.sqrt(functools.reduce(
+            operator.add, (getattr(self, component)**2
+                           for component, cls in self.attr_classes.items()
+                           if not issubclass(cls, Angle))))
+
+
+    def _operation(self, operator, other, reverse=False):
+        result = self.to_cartesian()._operation(operator, other, reverse)
```
Comment (MSeifert04) [inline astropy/coordinates/representation.py]: I'm not sure if it's a good idea to use a variable with the same name as a module you imported. Maybe `op` instead?

However this is not really important.

**Comment by MSeifert04** [comment]
I've left some stylistic comments. :+1:

File: astropy/coordinates/representation.py [hunk 3]
```diff
@@ -83,8 +84,13 @@ class BaseRepresentation(ShapedLikeNDArray):
     that frame classes may override this with their own preferred units.
     """
 
+    attr_classes = OrderedDict()
+
     recommended_units = {}  # subclasses can override
 
+    # Ensure multiplication/division with ndarray doesn't lead to object arrays.
+    __array_priority__ = 50000
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: Really all this needs to do is to ensure any `ndarray` subclass has lower priority. Of course, I don't know what the highest possible one is...

File: astropy/coordinates/representation.py (hunk 1 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: You're right that these could be unified. I now did and decided to use a plain loop rather than list comprehension, which should make it a bit easier to follow (and allows me to pass back a `NotImplemented` in the process, in case `other` is a filehandle or so...).

File: astropy/coordinates/representation.py (hunk 2 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: Done; agreed that it is good not to give causes for confusion!

**Comment by astrofrog** [comment]
I'm trying to wrap my head around the meaning of this for spherical frames. For a unit spherical frame, is the mean the circular mean, or can one end up with points that are floating inside the unit sphere? (in which case the mean of a unit spherical frame is not a unit spherical frame).

File: astropy/coordinates/representation.py [hunk 4]
```diff
@@ -221,6 +228,143 @@ def _unitstr(self):
...
+        return np.logical_not(self.__eq__(other))
+
+    def _scale_operation(self, op, other):
+        results = []
+        for component, cls in self.attr_classes.items():
+            value = getattr(self, component)
+            if issubclass(cls, Angle):
+                results.append(value)
+            else:
+                result = op(value, other)
```
Comment (astrofrog) [inline astropy/coordinates/representation.py]: Does this allow for instance a spherical representation to be multiplied by a temperature? Do we need to restrict ourselves to dimensionless quantities?

**Comment by astrofrog** [comment]
I'm not sure how I feel about adding to arithmetic to all different representations, since in some cases I think the meaning isn't super obvious (and there could be multiple interpretations - e.g. circular mean vs geometric mean). Just to put the idea out there, what about implementing this only in the cartesian representation class, which would then require an explicit conversion to/from cartesian?

File: astropy/coordinates/representation.py (hunk 4 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: Any representation can be multiplied by any quantity: it is just the ones that carry dimensional information (i.e., not the angles, so, e.g., r and z for cylindrical) that get multiplied.

**Comment by mhvk** [comment]
@astrofrog - originally, @StuartLittlefair and I mostly thought about operations on cartesian representations, and possible we should indeed limit ourselves to it. My main rationale for not doing so is that some operations are truly obvious in other frames as well (especially, `abs`, but also multiplication and division) and for the others the only easy-to-explain, unambiguous way of doing the arithmetic is going via cartesian, and one might as well make that the default. And once addition seems only logically done via cartesian, sum and mean follow... That said, my guess is that all but `abs`, `__mul__` and `__div__` would be rarely used for the non-cartesian representations. Furthermore, it is not completely obvious one would necessarily want the result transformed back to spherical.

For `UnitSphericalRepresentation`, essentially any operation on it makes the result not be on the unit sphere, so I always return a `SphericalRepresentation`. Obviously, a user could transform this back to the unit sphere if so desired.

**Comment by astrofrog** [comment]
I'm not sure I would say ``abs`` is unambiguous - if you have points in spherical polar coordinates, does it refer to the length of the vector, or does it make all values positive? (by analogy to how it works in the complex plane).

**Comment by astrofrog** [comment]
Division is also not unambiguous in spherical polar coordinates - if I divide by two, will it divide the angles by two as well? I think those operations only make sense in a cartesian frame - dividing spherical coordinates by two makes no conceptual sense.

**Comment by mhvk** [comment]
For complex numbers, `abs` returns the length of the vector, it does not make all values positive!
```
In [1]: abs(12-5j)
Out[1]: 13.0
```
For representations, it seems that to do the same is the only natural option, since representations really are just different ways of representing 3D vectors.

And thinking of vectors, **x**/|**x**| should be a unit vector, so that defines division pretty unambiguously too -- to me at least, it would be very weird to have longitudes or latitudes changed by division (not to mention that in that case the divisor would have to be dimensionless, so one could never get a velocity out by dividing by a quantity with time units).

**Comment by astrofrog** [comment]
> For complex numbers, abs returns the length of the vector, it does not make all values positive!

Yes sorry, that was my point - it wasn't obvious whether abs() would work like on a quantity, or if it would always give the length of the vector. If it'll always return the length of the vector, then that's fine.

**Comment by astrofrog** [comment]
I guess I'm just trying to make sure that things are truly unambiguous here for users - I agree that users will think of cartesian representation as a vector, but I'm not sure if the same is true for the others (not saying I don't agree, just that it took me a while to understand how this would work for non-cartesian frames). I guess part of me would be in favor or implementing this only on the cartesian representation initially then seeing later if there is a need to extend to others. But at the end of the day, I don't have too strong an opinion, so I'll leave it up to others.

I think @eteq should definitely have a say on this first, and @Cadair (who wrote the original classes with me)

**Comment by mhvk** [comment]
Agreed it is only logical to the extent one thinks of these classes all being representations of vectors. Looking forward to comments from others.

p.s. I realise I forgot `__neg__` among the methods that can be supported. Will add it still.

**Comment by mhvk** [comment]
Looking at other possible methods, the only one left that might be useful is `__matmul__` (and its reverse). For `CartesianRepresentation`, the reverse of that would logically be `matrix @ cart_rep`, so `__rmatmul__` could be directly mapped to `self.transform(matrix)`. Similarly, `cart_rep @ matrix` would effectively map to `self.transform(matrix.T)`.

For this one, though, I think I would not automatically extend it to `SphericalRepresentation`, at least as long as `matrix` would just be an `ndarray` with shape ending in 3,3, since it is not obvious that those have to be cartesian elements (unlike for addition with another representation, where one does know what form it is and can explicitly transform to cartesian).

Given this, and that this really only becomes important when our minimum python version is 3.5, I think I'll leave `__matmul__` for another PR.

**Comment by eteq** [comment]
I'm still mulling this over, but some thoughts right now:

The reason for not implementing `__eq__` in the first place was driven by the following example (or at least situations *like* this):
```python
>>> cr = CartesianRepresentation(np.random.randn(3,4)*u.kpc)
>>> sr = cr.represent_as(SphericalRepresentation)
>>> cr2 = sr.to_cartesian()
>>> cr.x == cr2.x, cr.y == cr2.y, cr.z == cr2.z
(array([False, False, False,  True], dtype=bool),
 array([False, False, False, False], dtype=bool),
 array([ True, False,  True,  True], dtype=bool))
```
So that means the following happens in *this* PR's:
```
>>> cr = CartesianRepresentation(np.random.randn(3,4)*u.kpc)
>>> sr = cr.represent_as(SphericalRepresentation)
>>> sr == cr 
array([ True,  True,  True,  True], dtype=bool)
>>> cr == sr 
array([False, False, False, False], dtype=bool)
```
Hence, it appears that equality is not commutative! Most users will be very surprised by this and not realize it until they've written buggy code.

The underlying problem here is that machine precision isn't sufficient to maintain perfect round-tripping, so this holds for pretty much any pair of representations to some extent or another.  Of course there are various workarounds for this, but to me (I think after being convinced by @taldcroft?), that then creates more problems than it solves.  So the idea was that we should decide for the user what "equality" means beyond the standard ``self is other`` equivalence.


I also echo @astrofrog's concern that most of these operations are *not* unambiguous for many of the representations - it might make sense to all of us in this conversation because we've thought about it a fair amount, but the vast majority of users won't, and they might make incorrect assumptions.  More abstractly, it's not clear that a representation really *should* be thought of as "generic vector".  As currently imagined they are *not*, rather they are specifically spatial locations. Adding all these operations starts blurring that.

So I'm going to through out two alternatives that still maintain much of the capability here while not making as many assumptions:

1. Create a `CartesianVector` that's a sub-class of cartesian representation, and it gets all the operations.  It's initializer should accept representations and just pass them through if it's cartesian (so low performance overhead), otherwise it does `to_cartesian`.  Maybe also add `to_cartesian_vector` or something like that to the base representation to make it easier. (I favor this one)
2. Move all the operations to `CartesianRepresentation`, and make it clear in the docs that that's intentional - basically a slightly-more-convenient (but also a bit more muddled) version of option 1.

**Comment by mhvk** [comment]
@eteq - I'll need to think a bit more about your points, but as you already discuss `__eq__` separately, think it is good to perhaps discuss the various methods separately as well, since I think for some I can see the arguments for them being only on Cartesian much more than the others. So, in order of increasing obscurity for non-cartesian:
1. `__abs__`: to me, the one that whose definition is most obvious for all representations.
2. `__mul__`, `__div__`, `__neg__`, `__pos__`: I feel these are very well-defined, since they only make sense if they apply to the unit-ful components; and getting a normalised version out by dividing by `abs()` seems one of those simply useful things.
3. `dot()`, `cross()`: clearly defined in cartesean, arguably very handy to at least allow the second element to be non-cartesian, so one can project on `UnitSpherical`.
4. `__add__`, `__sub__`: these are obvious for cartesian, less so for the others. Possibly nice to subtract two spherical coordinates and get a properly directed proper motion out.
5. `mean()`, `sum()`: follow on from `__add__`, really.
6. `__eq__`: it was there because it was easy to do and helps writing test cases, but I agree that in practice it may be less than useful for any of them.

So, one alternative would be to keep (1) and (2) for all (those are also the ones that do not assume cartesian in their implementation), and move 3,4,5 to Cartesian proper (can always add them later to the others, but hard to remove).

But I'll also argue for just keeping things as I proposed (except perhaps remove `__eq__`): these are *not* meant for regular users -- those only see `SkyCoord` and I am not proposing at all that we provide arithmetic for those. These methods are meant to make future development of coordinates easier, and allow broader use of the representations as a vector.

For coordinates, I see particular use & need once we start implementing proper motions and velocities: we will need simple ways to add and subtract vectors, divide and multiply by time spans, and calculate cross and dot products. I think methods are better there than helper functions (indeed, a good example is `transform_to`, which used to be a helper function). Also, I do think at that point we want to be able to use the various representations to represent velocity vectors.

As for broader use, I have in fact a `QuantityVector` class PR sitting in my git locally, which has the benefit of allowing arbitrary dimension, but also made me realise that this is not a logical `Quantity` subclass, as too many methods and `ufunc`s just don't make sense on vectors. This PR solves that conundrum for me. 

Now, arguably, that would be solved just as nicely with a `CartesianVector` class, but then why prevent me from transforming it to spherical coordinates if I like that?

Overall, it may boil down to a choice: I *do* think representations are really just different ways to represent arbitrary 3-D vectors (with arbitrary units).

**Comment by mhvk** [comment]
On `__eq__`, one reason for having it is that it would be nice if `rep[0] == rep[0]` would yield `True` (and similarly for other slices, getting their overlap, etc.). But your argument is strong too... One could simply not convert to any other representation. Or just not implement it. Anyway, not the most important of the things to decide!

**Comment by taldcroft** [comment]
I read through this thread and started off on the ambivalent side of the "err, seems a bit ambiguous, etc".  But as I was reading I started thinking about the two ways we can conceptually think of representations:

1. Representation of an abstract point in a frame
2. Representation of an abstract vector in a frame from the origin to the point

From the code and math perspective I think these are equivalent [??], but if you fully adopt idea (1) then basically none of the operations really make sense (what does it mean to add two points in space?).  BUT if you re-orient yourself to idea (2) then I think all of the operations make sense for all representations and you gain powerful tools.  For instance, a useful equality test for two representations could be something like:
```
np.all(abs(r1 - r2) < 1e-7) 
```

So by the end I was convinced that we should accept all the operators on all representations except for `__eq__`, which is basically not going to work.  

So @eteq suggested making the `Vector` bit explicit with a subclass, but I think that at the end of the day this would just end up creating confusion and duplication where it really isn't necessary.  I.e. users would ask "when should I use `CartesianRepresentation` and when should I use `CartesianVector`?"  The answer would be "use the vector if you might want to use this list of operators on it later".  But there is nothing code wise that prevents doing this all in the `*Representation` classes, it is just conceptual.

Now I confess that I haven't gone back through the docs to see how representations are presented, but perhaps some/many users already think of the abstract thing underneath as a vector rather than a point?

The only real outstanding question to me is how to handle in general the question of type promotion for allowed operations.  For instance we know that the output of most operations with `UnitSpherical` needs to be promoted to a representation that isn't Unit.  In theory user-defined representation classes could have other constraints, so a generalized mechanism would be good.  This is certainly a solvable problem.

**Comment by mhvk** [comment]
OK, I think there is quite clear consensus that we should not have `__eq__`, so I removed that (and rebased as well, so this remains mergeable).

**Comment by adrn** [comment]
@mhvk Sorry for the delay, but I'd like to take a close look at this over the weekend.

File: astropy/coordinates/representation.py [hunk 5]
```diff
@@ -400,6 +529,85 @@ def transform(self, matrix):
...
+
+        Parameters
+        ----------
+        other : `~astropy.coordinates.BaseRepresentation`
+            If not already cartesian, it is converted.
+
+        Returns
+        -------
+        dot_product : `~astropy.units.Quantity`
+            The sum of the product of the x, y, and z components of ``self``
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: I seem to be blind to what actually is wrong here... Could you elucidate?

File: astropy/coordinates/representation.py [hunk 6]
```diff
@@ -83,8 +84,14 @@ class BaseRepresentation(ShapedLikeNDArray):
     that frame classes may override this with their own preferred units.
     """
 
+    attr_classes = OrderedDict()
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: Good question. Removing it doesn't seem to matter, so I did. I think it may be a leftover from an earlier attempt.

File: astropy/coordinates/representation.py [hunk 7]
```diff
@@ -221,6 +228,128 @@ def _unitstr(self):
...
+        return self._combine_operation(operator.sub, other, reverse=True)
+
+    def mean(self, *args, **kwargs):
+        """Return the vector mean along a given axis.
+
+        The representation is converted to cartesian, the means of the x, y,
+        and z components are calculated, and the result is converted back to
+        the same representation as the input.
+
+        Refer to `numpy.mean` for full documentation of the arguments.
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: OK, done

File: CHANGES.rst [hunk 8]
```diff
@@ -20,6 +20,11 @@ New Features
     For JPL kernels, this roughly doubles the execution time, so if one requires
     only the positions, one should use ``get_body_barycentric``. [#5231]
 
+  - Arithmetic on representations, such as addition of two representations,
```
Comment (mhvk) [inline CHANGES.rst]: Good point. Done.

File: CHANGES.rst [hunk 9]
```diff
@@ -20,6 +20,11 @@ New Features
...
+    multiplication with a ``Quantity``, or calculating the norm via ``abs``,
+    has now become possible. Furthermore, there are news methods ``mean``,
```
Comment (mhvk) [inline CHANGES.rst]: OK, done

File: astropy/coordinates/representation.py [hunk 10]
```diff
@@ -421,6 +629,7 @@ class UnitSphericalRepresentation(BaseRepresentation):
     attr_classes = OrderedDict([('lon', Longitude),
                                 ('lat', Latitude)])
     recommended_units = {'lon': u.deg, 'lat': u.deg}
+    _dimensional_representation = None  # set to SphericalRepresentation below
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: Yes, much nicer. Done.

**Comment by mhvk** [comment]
@taldcroft - I address your comments except that for one I wasn't quite sure what you meant with a "copy and paste" error (maybe I addressed it, but please check).

Also rebased, since this was getting a bit far behind. I think I'll call this ready for final review now...

File: astropy/coordinates/representation.py (hunk 5 above)
Comment (taldcroft) [inline astropy/coordinates/representation.py]: Ah sorry, I was being blind and did not see the "of the product" words.  So never mind.

File: astropy/coordinates/representation.py [hunk 11]
```diff
@@ -221,6 +226,138 @@ def _unitstr(self):
...
+
+    def __neg__(self):
+        return self._scale_operation(operator.neg)
+
+    # Follow numpy convention and make an independent copy.
+    def __pos__(self):
+        return self.__class__(*(getattr(self, component)
+                                for component in self.components), copy=True)
+
+    def __abs__(self):
```
Comment (adrn) [inline astropy/coordinates/representation.py]: If we keep this, I'm also curious here why you use `functools` and not:
```python
return np.sqrt(np.sum([getattr(a, component)**2
                           for component in a.components], axis=0))
```
(which is a bit faster I think)

File: astropy/coordinates/tests/test_representation_arithmetic.py [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+    def test_abs(self, representation):
+        in_rep = self.spherical.represent_as(representation)
+        abs_rep = abs(in_rep)
+        assert isinstance(abs_rep, u.Quantity)
+        assert_quantity_allclose(abs_rep, self.distance)
+
+    def test_abs_unitspherical(self):
+        abs_rep = abs(self.unit_spherical)
+        assert abs_rep.unit == u.dimensionless_unscaled
+        assert np.all(abs_rep == 1. * u.dimensionless_unscaled)
```
Comment (adrn) [inline astropy/coordinates/tests/test_representation_arithmetic.py]: I'm surprised this doesn't need an `allclose()`, but ok

**Comment by adrn** [comment]
@mhvk The rest of the tests look ok, but see comments above. Thanks for your work on this!

File: astropy/coordinates/representation.py (hunk 11 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: The route via `numpy` is in fact substantially slower, mostly because of the creation of the new array. E.g.,
```
sr = SphericalRepresentation(0.*u.deg, 45*u.deg, np.arange(100000.)*u.km)
cr = sr.to_cartesian()
# check Cartesian
%timeit abs(cr)
# 100 loops, best of 3: 3.4 ms per loop
%timeit np.sqrt(np.sum([getattr(cr, component)**2 for component, cls in cr.attr_classes.items() if not issubclass(cls, Angle)], axis=0))
# 100 loops, best of 3: 6.49 ms per loop
# now for Spherical
%timeit abs(sr)
# 1000000 loops, best of 3: 520 ns per loop
%timeit np.sqrt(np.sum([getattr(sr, component)**2 for component, cls in sr.attr_classes.items() if not issubclass(cls, Angle)], axis=0))
# 1000 loops, best of 3: 1.51 ms per loop
```

Comment (mhvk) [inline astropy/coordinates/representation.py]: Actually, checking actual values, more important is that this way the components are not mishandled: the `Quantity` components do not survive being in a list...

File: astropy/coordinates/representation.py [hunk 13]
```diff
@@ -400,6 +537,85 @@ def transform(self, matrix):
...
+            The sum of the product of the x, y, and z components of ``self``
+            and ``other``.
+        """
+        try:
+            other_c = other.to_cartesian()
+        except Exception:
+            raise TypeError("cannot only take dot product with another "
+                            "representation, not a {0} instance."
+                            .format(type(other)))
+        return functools.reduce(operator.add,
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: See above: the implicit conversion to an array of a list of `Quantity` does not work...
```
np.sum([10*u.m, 10*u.km])
ValueError: setting an array element with a sequence.
```

File: astropy/coordinates/tests/test_representation_arithmetic.py [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+    return functools.reduce(np.logical_and,
+                            (getattr(first, component) ==
+                             getattr(second, component)
+                             for component in first.components))
+
+class TestArithmetic():
+    """Test absolute value, multiplication, and division."""
+
+    def setup(self):
+        self.lon = Longitude(np.arange(0, 12.1, 2), u.hourangle)
```
Comment (mhvk) [inline astropy/coordinates/tests/test_representation_arithmetic.py]: The numbers are not really arbitrary, they ensure that various combinations give integer distances... I'll add a note to that effect.

File: astropy/coordinates/tests/test_representation_arithmetic.py (hunk 12 above)
Comment (mhvk) [inline astropy/coordinates/tests/test_representation_arithmetic.py]: For unit spherical, the length of the vectors should be identically unity, not approximately.

File: astropy/coordinates/representation.py (hunk 13 above)
Comment (adrn) [inline astropy/coordinates/representation.py]: Got it!

**Comment by mhvk** [comment]
@adrn - on your larger comments: I think the main confusion arises because, as commented by @taldcroft's (https://github.com/astropy/astropy/pull/5301#issuecomment-245408113), one can think of the representations as integrally part of coordinates, or as vectors from the origin. Only in the latter interpretation does my PR make much sense: it makes these representations very useful outside of `SkyCoord`, effectively giving us a `Vector` class that allows vector arithmetic as well as representing vectors in non-cartesian ways. (Just to be sure: I do not propose to propagate any of these methods up, although I think they will be useful internally to calculate proper motions, etc.)

If you think of these as vectors, then `abs(rep)` is unambiguous (just like `abs(complex)` is unambiguous, it just represents `|x|`), and similarly dividing or multiplying by a quantity makes sense (e.g., given some positions `pos1` and `pos2` and a quantity with time units `dt`, `vel = (pos1-pos2)/dt` gives a vector velocity).

**Comment by taldcroft** [comment]
On the topic of the name `abs` or `norm`, I agree with @adrn that `norm` is more intuitive and less prone to ambiguity.  With `abs`, even though it would not make practical or mathematical sense, some people might guess that it just takes that absolute value of the components.

Indeed in https://en.wikipedia.org/wiki/Absolute_value#Vector_spaces we read *"‖F‖ is called an absolute value, but more usually a norm, if it satisfies the following axioms:"*.

**Comment by mhvk** [comment]
I'd like to keep `abs` since I don't feel one should decide based on what people might do wrong with it. That said, I'm quite happy to add a `norm()` method. Should it, like `np.linalg.norm` have an `ord` parameter?

**Comment by mhvk** [comment]
Just to get a sense: do people feel this should go in? If so, I will add a new documentation page, maybe linked both from coordinates and units, which makes clear that this is a way to use representations for general vector purposes.

**Comment by StuartLittlefair** [comment]
@mhvk I'll raise my voice in favour. I am looking forward to being able to use this code to simplify a lot of the coordinate transformation framework.

**Comment by taldcroft** [comment]
@mhvk - well there is the "one and preferably only one way to do it".  That said I won't argue too strongly with including `abs` in addition to `norm`.  Overall definitely :+1: on this.

**Comment by taldcroft** [comment]
After walking away and thinking about the "one and only one way", I remembered the built-in `abs` function and starting wavering on having a `norm` method at all.  Sigh, indecision.

**Comment by pllim** [comment]
There can be only one

**Comment by mhvk** [comment]
OK, I added some documentation in the "using and designing coordinate representations" section, with a link to it from quantity. For now, I did not implement `norm`; my sense is that we can do this later if needed (maybe together with `__matmul__`...).

**Comment by mhvk** [comment]
See https://github.com/astropy/astropy/pull/5273/files/de00e2ee92d85039425a6224994e2e3f5fe44ed9#r83705780 for an example how this would help the coordinate transformation code...

**Comment by taldcroft** [comment]
I'm good with this now.  

@adrn @eteq - do you plan to do further review?

**Comment by mhvk** [comment]
Rebased now that #5254 is in.

**Comment by adrn** [comment]
I'm happy with this -- ok to merge 👍

**Comment by eteq** [comment]
FYI, I'm planning to do a review on this shortly.

**Comment by eteq** [comment]
I also found two other problems while experimenting - I created mhvk/astropy#20 against @mhvk's branch with the regression tests for these:

1. `mul` fails if the shape of the representation is not 1d.  Some of the others do too, but I *think* they're all due to `mul` failing?  Anyway, you can see in the tests @mhvk and maybe you can tell offhand where the problem is?
2. `abs` does not give consistent types - sometimes its a `Quantity` and sometimes a `Distance`.  But in either case its a numpy array of object type, not a Quantity/Distance with the appropriate shape

File: astropy/coordinates/representation.py [hunk 15]
```diff
@@ -230,6 +234,140 @@ def _unitstr(self):
...
+        return self._scale_operation(operator.neg)
+
+    # Follow numpy convention and make an independent copy.
+    def __pos__(self):
+        return self.__class__(*(getattr(self, component)
+                                for component in self.components), copy=True)
+
+    def __abs__(self):
+        return np.sqrt(functools.reduce(
+            operator.add, (getattr(self, component)**2
```
Comment (StuartLittlefair) [inline astropy/coordinates/representation.py]: I think we should avoid conversions to ```Distance``` - one might want to use ```CartesianRepresentations``` to represent velocities for example.

**Comment by mhvk** [comment]
@eteq - thanks for the comments. I'm really confused about the failures, though: for me, both multiplication and `abs` work perfectly fine on multi-d data! Could you check that you really do get a problem with getting object `dtype`? (as pointed out by @StuartLittlefair, `abs` can and should sometimes return `Quantity` rather than `Distance` -- it only returns `Distance` for spherical coordinates with length units because for those the radius is initialized as a distance).

File: astropy/coordinates/representation.py (hunk 15 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: `self.dot` is slower for representations other than `Cartesian`, since it first converts to `Cartesian` and then takes the product. Although this really only matters for `Cylindrical` since for the spherical ones I already override it. Still, my sense is to leave it as is.

File: astropy/coordinates/representation.py [hunk 16]
```diff
@@ -230,6 +234,140 @@ def _unitstr(self):
...
+
+        Averaging is done by converting the representation to cartesian, and
+        taking the mean of the x, y, and z components. The result is converted
+        back to the same representation as the input.
+
+        Refer to `~numpy.mean` for full documentation of the arguments.
+
+        Returns
+        -------
+        mean : `~astropy.coordinates.BaseRepresentation`
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: Good point, will do.

File: astropy/coordinates/representation.py [hunk 17]
```diff
@@ -409,6 +547,85 @@ def transform(self, matrix):
...
+            return NotImplemented
+
+        first, second = ((self, other_c) if not reverse else
+                         (other_c, self))
+        return self.__class__(*(op(getattr(first, component),
+                                   getattr(second, component))
+                                for component in first.components))
+
+    def mean(self, *args, **kwargs):
+        """Return the vector mean along a given axis.
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: Again a good point.  I removed "along the given axis" as I agree it is only confusing, and expanded text as follows:
```
        """Vector mean.

        Averaging is done by converting the representation to cartesian, and
        taking the mean of the x, y, and z components. The result is converted
        back to the same representation as the input.

        Refer to `~numpy.mean` for full documentation of the arguments, noting
        that ``axis`` is the entry in the ``shape`` of the representation, and
        that the ``out`` argument cannot be used.
```

File: docs/coordinates/representations.rst [hunk 18]
```diff
@@ -191,3 +205,73 @@ A representation class may also have a ``_unit_representation`` attribute
 probably only meaningful for subclasses of
 `~astropy.coordinates.SphericalRepresentation`, where it is assumed that it
 will be a subclass of `~astropy.coordinates.UnitSphericalRepresentation`.
+
+
+Vector arithmetic
+^^^^^^^^^^^^^^^^^
+
+Representations support basic vector arithmetic, in particular taking the
```
Comment (mhvk) [inline docs/coordinates/representations.rst]: I tried to do this with the note at the very end, but now agree this would come too late. I thought a bit about changing the text, but ended up deciding to move the note higher up and make it a proper `.. Note::`, so it stands out in the documentation.

**Comment by mhvk** [comment]
On `__abs__` -- I still find it quite obvious, but as you are the third to disagree, I've now changed it to a `norm` method. I do see that it is an obvious choice as well, and an additional advantage is that it is easier to find. (I removed `__abs__` altogether as I don't want anybody ever to take just absolute values of the components...)

**Comment by eteq** [comment]
> for me, both multiplication and abs work perfectly fine on multi-d data! 

Hmm, so if I add eteq/astropy@ec9cba84a040d0b6004891ea1773438b4b5946ec (from mhvk/astropy#20) in on top of this branch as it stands right now, I get ``ValueError: operands could not be broadcast together with shapes (4,7) (3,1)``  failures when I try it locally.  But if I take the tests as they stand in the branch *right now*, they all pass.   I think  this is because the problem is in `__mul__`, and right now it looks like `__mul__` (and at least some of the other operations) don't get tested in more than 1D?  That's why I added that test in eteq/astropy@ec9cba84a040d0b6004891ea1773438b4b5946ec, to make sure *everything* gets hit by the test suite in the n-D case.

Are you seeing it pass even if you try  eteq/astropy@ec9cba84a040d0b6004891ea1773438b4b5946ec ?

> Could you check that you really do get a problem with getting object `dtype`?

That, however, seems to have disappeared in the latest version, so all good there. 

>  `abs` can and should sometimes return `Quantity` rather than `Distance`

Ah, good point!  I get it now.  I can see arguments one way or another for whether it should be forced either way, but I think it's probably best to leave it alone, as it's not clear which is better (and it's all `Quantity`-compatible anyway).

**Comment by mhvk** [comment]
@eteq - yes, I tried your examples by hand, and they work on my machine. But just to be sure, in the latest commits I added a test (for `__div__`, but they use the same code path) on a 2-D array that the `cross` test had created anyway (see https://github.com/astropy/astropy/pull/5301/files#diff-3d38462c8d8bf432919ce1decd35579bR345) and that new tests also passes locally (sadly, don't know about travis!). Does the current PR pass for you as well? If it fails, I can only think it must be a python2 vs 3 issue; what python are you on?

**Comment by mhvk** [comment]
@eteq - the new tests, including the 2-D array check, pass on travis for all settings. So, I went ahead and merged as I would like to rebase the usage to get that shipshape as well. If there still are problems with this PR, I can address them as part of #5434 (which I'll leave open until you're OK).

**Comment by eteq** [comment]
@mhvk - hmm, I'm still seeing this error in both py2 *and* py3.  I'll make a PR into astropy just adding the test I described and lets see what travis tells us.  That way we'll at least know if its local to my computer somehow, and can try to fix the error in that PR if so.

**Comment by taldcroft** [comment]
:tada:
